Optimize IOrderedEnumerable.Take(1)

AndreyAkinshin opened the pull request:
Calls like `array.OrderBy(x => x).Take(1)` use `EnumerableSorter.ElementAt` that invokes `QuickSelect`. `Take(1)` is just the Minimum for `OrderedEnumerable` and can implemented more efficiently. The Minimum-based implementation has a big impact on the worst case time complexity: now it's `O(N)` instead of `O(N^2)`.

The performance test results:

```
Before:

| Metric   | Unit | Iterations |  Average | STDEV.S |      Min |      Max
|:-------- |:----:|:----------:| --------:| -------:| --------:| --------:
| Duration | msec |     2      | 8913.304 |  69.896 | 8863.880 | 8962.728

After:

| Metric   | Unit | Iterations | Average | STDEV.S |   Min |   Max
|:-------- |:----:|:----------:| -------:| -------:| -----:| -----:
| Duration | msec |    1000    |   2.765 |   0.330 | 2.555 | 6.162
```

The BenchmarkDotNet-based benchmark:

```cs
public enum Pattern
{
    QsWorst, // 2 4 6 8 10 5 3 7 1 9 (Worst QuickSort case)
    Equal, // 0 0 0 0 0 0 0 0 0 0
    Sorted, //  0 1 2 3 4 5 6 7 8 9
    Reversed, // 10 9 8 7 6 5 4 3 2 1
    Shuffled
}

public static class PatternGenerator
{
    private static int OddPos(int n, int k)
    {
        var s = k;
        while (s * 2 < n)
            s *= 2;
        return (n - 1) % s + 1;
    }

    private static int[] QsWorst(int n)
    {
        var a = new int[n];
        for (int x = 1; x <= n; x++)
        {
            if (x % 2 == 0)
                a[x / 2 - 1] = x;
            else if (x == n)
                a[n / 2] = x;
            else
                a[n - OddPos(n, x)] = x;
        }

        return a;
    }

    public static int[] Generate(int n, Pattern pattern)
    {
        switch (pattern)
        {
            case Pattern.QsWorst:
                return QsWorst(n);
            case Pattern.Equal:
                return Enumerable.Range(0, n).Select(x => 0).ToArray();
            case Pattern.Sorted:
                return Enumerable.Range(0, n).Select(x => x).ToArray();
            case Pattern.Reversed:
                return Enumerable.Range(0, n).Select(x => n - x).ToArray();
            case Pattern.Shuffled:
            {
                var rnd = new Random(42);
                return Enumerable.Range(0, n).Select(x => rnd.Next()).ToArray();
            }
            default:
                throw new ArgumentOutOfRangeException(nameof(pattern), pattern, null);
        }
    }
}


public class TakeBenchmark
{
    private int[] array;

    [Params(1000, 10000)] public int N;

    [Params(Pattern.QsWorst, Pattern.Equal, Pattern.Sorted,
            Pattern.Reversed, Pattern.Shuffled)]
    public Pattern Case;

    [GlobalSetup]
    public void Setup() => array = PatternGenerator.Generate(N, Case);

    [Benchmark(Description = "Take(1)")]
    public int Take()
    {
        var sum = 0;
        foreach (var x in array.OrderBy(x => x).Take(1))
            sum += x;
        return sum;
    }
}

class Program
{
    static void Main() => BenchmarkRunner.Run<TakeBenchmark>();
}
```

Before:

|     N |     Case |          Mean |          Error |        StdDev |
|------ |--------- |--------------:|---------------:|--------------:|
|  1000 |    Equal |      42.16 us |      0.4111 us |     0.1068 us |
|  1000 |  QsWorst |   8,853.04 us |    326.7693 us |    84.8771 us |
|  1000 | Reversed |      44.25 us |      2.2270 us |     0.5785 us |
|  1000 | Shuffled |      41.48 us |      0.2996 us |     0.0778 us |
|  1000 |   Sorted |      41.43 us |      0.4115 us |     0.1069 us |
| 10000 |    Equal |     415.93 us |      3.6485 us |     0.9477 us |
| 10000 |  QsWorst | 876,433.01 us | 19,096.6708 us | 4,960.2892 us |
| 10000 | Reversed |     441.04 us |     11.7266 us |     3.0459 us |
| 10000 | Shuffled |     298.97 us |      5.9536 us |     1.5464 us |
| 10000 |   Sorted |     412.12 us |      3.3257 us |     0.8638 us |

After:

|     N |     Case |      Mean |      Error |    StdDev |
|------ |--------- |----------:|-----------:|----------:|
|  1000 |    Equal |  25.50 us |  3.5458 us | 0.9210 us |
|  1000 |  QsWorst |  25.94 us |  2.1703 us | 0.5637 us |
|  1000 | Reversed |  25.54 us |  0.9774 us | 0.2539 us |
|  1000 | Shuffled |  24.39 us |  0.6551 us | 0.1702 us |
|  1000 |   Sorted |  24.23 us |  0.4637 us | 0.1205 us |
| 10000 |    Equal | 250.48 us | 18.0682 us | 4.6931 us |
| 10000 |  QsWorst | 244.50 us |  1.8494 us | 0.4804 us |
| 10000 | Reversed | 250.53 us | 30.2642 us | 7.8610 us |
| 10000 | Shuffled | 243.64 us | 16.0393 us | 4.1662 us |
| 10000 |   Sorted | 241.09 us |  2.6659 us | 0.6925 us |

danmosemsft commented:
Thanks. Are there sufficient correctness tests already?

AndreyAkinshin commented:
@danmosemsft, it seems that we don't have such tests. Should I add them in `TakeTests.cs`/`TakeLastTests.cs` or `OrderByTests.cs`?

danmosemsft commented:
@andreyakinshin whichever you prefer. Thanks!

danmosemsft commented:
Your benchmark does not seem to cover Take Last. I guess that is self evidently the same though

danmosemsft commented:
You might also consider whether it is worth adding your performance benchmark into the repo for regression protection.

AndreyAkinshin commented:
@danmosemsft, I updated the PR. Now it contains the correctness tests and performance tests. It turned out that `array.OrderBy(x => x).TakeLast(1)` doesn't use `QuickSelect`, so I removed the corresponding logic. In this PR, only `Take(1)` is optimized.

stephentoub commented:
Unless I'm missing it, I only see tests here for Take(1), which is the case this optimization is meant to improve.  What about other N values, where this optimization won't help?  Does it hurt at all?

I'm not against the optimization if there's no measurable impact on other cases; I just question how valuable this actually is... after all, a developer doing OrderBy(...).Take(1) could more easily write Min(...) themselves.

stephentoub commented on src/System.Linq/src/System/Linq/OrderedEnumerable.cs:
```diff
@@ -281,8 +281,13 @@ internal int[] Sort(TElement[] elements, int count, int minIdx, int maxIdx)
-            elements[QuickSelect(ComputeMap(elements, count), count - 1, idx)];
+        internal TElement ElementAt(TElement[] elements, int count, int idx)
+        {
+            var map = ComputeMap(elements, count);
```
Nit: var => int

stephentoub commented on src/System.Linq/src/System/Linq/OrderedEnumerable.cs:
```diff
@@ -281,8 +281,13 @@ internal int[] Sort(TElement[] elements, int count, int minIdx, int maxIdx)
+            var map = ComputeMap(elements, count);
+            if (idx == 0)
+                return elements[Min(map, count)];
+            return elements[QuickSelect(map, count - 1, idx)];
```
Nit: please use braces or a ternary, e.g.
```C#
if (idx == 0)
{
    return elements[Min(map, count)];
}
else
{
    return elements[QuickSelect(map, count - 1, idx)];
}
```
or:
```C#
return idx == 0 ?
    elements[Min(map, count)] :
    elements[QuickSelect(map, count - 1, idx)];
```

stephentoub commented on src/System.Linq/src/System/Linq/OrderedEnumerable.cs:
```diff
@@ -489,5 +496,14 @@ protected override int QuickSelect(int[] map, int right, int idx)
+            int index = 0;
+            for (int i = 1; i < count; i++)
+                if (CompareKeys(map[i], map[index]) < 0)
+                    index = i;
```
Nit: please use braces for both the for and if blocks

stephentoub commented on src/System.Linq/src/System/Linq/OrderedEnumerable.cs:
```diff
@@ -489,5 +496,14 @@ protected override int QuickSelect(int[] map, int right, int idx)
+        {
+            int index = 0;
+            for (int i = 1; i < count; i++)
+                if (CompareKeys(map[i], map[index]) < 0)
```
If two elements are equal according to CompareKeys, does this guarantee to always return the same element that OrderBy(...).Take(1) previously would?

AndreyAkinshin replied:
Internally, `CompareKeys` returns `index1 - index2` for equal objects to ensure the sorting stability. Thus, we will always be consistent with the old behavior.

stephentoub commented on src/System.Linq/tests/OrderByTests.cs:
```diff
@@ -422,5 +422,19 @@ public void SortsRandomizedEnumerableCorrectly(int items)
+        [InlineData(new[] { 2, 4, 6, 8, 10, 5, 3, 7, 1, 9 })]
+        public void TakeOne(IEnumerable<int> source)
+        {
+            foreach (var x in source.OrderBy(i => i).Take(1))
```
Nit: var => int

stephentoub commented on src/System.Linq/tests/Performance/Perf.Linq.cs:
```diff
@@ -42,6 +42,29 @@ public static IEnumerable<object[]> IterationSizeWrapperDataNoWrapper()
+        {
+            int OddPos(int k)
+            {
+                var s = k;
```
Nit: in general, please search and replace var with the proper type anywhere it's not explicitly stated on the RHS

stephentoub commented on src/System.Linq/tests/Performance/Perf.Linq.cs:
```diff
@@ -42,6 +42,29 @@ public static IEnumerable<object[]> IterationSizeWrapperDataNoWrapper()
                     yield return new object[] { size, iteration };
         }
 
+        private static int[] QuiskSortWorstCase(int n)
```
Nit: Quisk => Quick

stephentoub commented on src/System.Linq/tests/Performance/Perf.Linq.cs:
```diff
@@ -42,6 +42,29 @@ public static IEnumerable<object[]> IterationSizeWrapperDataNoWrapper()
+                else if (x == n)
+                    a[n / 2] = x;
+                else
+                    a[n - OddPos(x)] = x;
```
Nit: please use braces for such cascading if/else/else ifs.

AndreyAkinshin commented:
@stephentoub, Originally, I tried to optimize the `Take` method in general with the help of the partial heapsort. It has an awesome performance boost in the worst case, but it slows down the average case, so we need a better solution.
During benchmarking, I found out that `Take(1)` uses `QuickSelect` instead of `QuickSort` (we handle the case of a single element [here](https://github.com/dotnet/corefx/blob/v2.2.0-preview2/src/System.Linq/src/System/Linq/OrderedEnumerable.cs#L90)) I don't know why this case is so special: it seems that we can use `QuickSort` for `ElementAt` as well, it works the same way with the same time complexity.
Anyway, I already had a proven optimization for this small case (without any side-effects or impact on other cases), so I decided to create a pull request. In real life, such construction can be used as a special case of `Take(K)`.

stephentoub commented on src/System.Linq/tests/OrderByTests.cs:
```diff
@@ -422,5 +422,19 @@ public void SortsRandomizedEnumerableCorrectly(int items)
+        public void TakeOne(IEnumerable<int> source)
+        {
+            foreach (int x in source.OrderBy(i => i).Take(1))
+                Assert.Equal(source.Min(), x);
```
Are there tests elsewhere that validate OrderBy(...).Take(1) returns 1 and only 1 item if there's at least one item in the input?  Since we're modifying those code paths, it'd be good to validate that here, too, e.g.
```C#
int count = 0;
foreach (int x in source.OrderBy(i => i).Take(1))
{
    Assert.Equal(source.Min(), x);
}
Assert.Equal(1, count);
```

AndreyAkinshin replied:
Fixed.

stephentoub commented on src/System.Linq/src/System/Linq/OrderedEnumerable.cs:
```diff
@@ -281,8 +281,13 @@ internal int[] Sort(TElement[] elements, int count, int minIdx, int maxIdx)
+        internal TElement ElementAt(TElement[] elements, int count, int idx)
+        {
+            int[] map = ComputeMap(elements, count);
+            return idx == 0 ?
```
It's unlikely to affect anything measurably, but in the name of completeness, you've validated that this extra check / the extra code in the method doesn't negatively impact cases beyond the positive cases you created benchmarks for?

AndreyAkinshin commented:
@stephentoub 
> It's unlikely to affect anything measurably, but in the name of completeness, you've validated that this extra check / the extra code in the method doesn't negatively impact cases beyond the positive cases you created benchmarks for?

I improved the Welch's t-test in BenchmarkDotNet, now it's possible to specify a threshold for distribution comparing (the update is available in v0.11.1.754-nightly; it will be included in v0.11.2). Here is my benchmark:

```cs
public class MyConfig : ManualConfig
{
    public MyConfig()
    {
        var baseJob = Job.Default
            .WithWarmupCount(10)
            .WithIterationCount(300)
            .WithInvocationCount(2097152)
            .WithOutlierMode(OutlierMode.None);
        Add(baseJob.With(new CoreRunToolchain(new FileInfo(
                "/home/akinshin/Work/dotnet/corefx-upstream/bin/runtime/netcoreapp-Linux-Debug-x64/corerun")))
            .WithId("c8e3ba6c09")
            .AsBaseline()
        );
        Add(baseJob.With(new CoreRunToolchain(new FileInfo(
                "/home/akinshin/Work/dotnet/corefx/bin/runtime/netcoreapp-Linux-Debug-x64/corerun")))
            .WithId("corefx#32389")
        );
    }
}

[Config(typeof(MyConfig))]
[WelchTTestAbsoluteNanoseconds(0.3)]
[RPlotExporter]
public class TakeOverheadBenchmark
{
    private int[] array = {1, 2, 3};

    [Benchmark]
    public int Take1()
    {
        int sum = 0;
        foreach (var x in array.OrderBy(x => x).Take(1))
            sum += x;
        return sum;
    }

    [Benchmark]
    public int Take2()
    {
        int sum = 0;
        foreach (var x in array.OrderBy(x => x).Take(2))
            sum += x;
        return sum;
    }
}
```
The `corefx-upstream` folder contains the `c8e3ba6c09` revision; the `corefx` folder contains `c8e3ba6c09` plus the patch from this PR. The iteration count is 300 to provide better statistics metrics for distributions. The invocation count is fixed (`2097152`) to provide the same accuracy between different runtimes. Here are my results:

```

BenchmarkDotNet=v0.11.1.754-nightly, OS=ubuntu 18.04
Intel Core i7-7700K CPU 4.20GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.402
  [Host]       : .NET Core 2.1.4 (CoreCLR 4.6.26814.03, CoreFX 4.6.26814.02), 64bit RyuJIT
  c8e3ba6c09   : .NET Core ? (CoreCLR 4.6.26920.03, CoreFX 4.6.27019.0), 64bit RyuJIT
  corefx#32389 : .NET Core ? (CoreCLR 4.6.26920.03, CoreFX 4.6.27019.0), 64bit RyuJIT

OutlierMode=None  Toolchain=CoreRun  InvocationCount=2097152  
IterationCount=300  WarmupCount=10  

 Method |          Job |     Mean |     Error |   StdDev |   Median | Scaled | ScaledSD | p-value |
------- |------------- |---------:|----------:|---------:|---------:|-------:|---------:|--------:|
  Take1 |   c8e3ba6c09 | 325.1 ns | 0.7870 ns | 4.101 ns | 323.7 ns |   1.00 |     0.00 |  1.0000 |
  Take1 | corefx#32389 | 315.5 ns | 0.4522 ns | 2.357 ns | 314.7 ns |   0.97 |     0.01 |  0.0000 |
        |              |          |           |          |          |        |          |         |
  Take2 |   c8e3ba6c09 | 350.3 ns | 0.7037 ns | 3.667 ns | 349.1 ns |   1.00 |     0.00 |  1.0000 |
  Take2 | corefx#32389 | 348.6 ns | 0.9457 ns | 4.929 ns | 347.2 ns |   1.00 |     0.02 |  1.0000 |

// * Legends *
  Mean     : Arithmetic mean of all measurements
  Error    : Half of 99.9% confidence interval
  StdDev   : Standard deviation of all measurements
  Median   : Value separating the higher half of all measurements (50th percentile)
  Scaled   : Mean(CurrentBenchmark) / Mean(BaselineBenchmark)
  ScaledSD : Standard deviation of ratio of distribution of [CurrentBenchmark] and [BaselineBenchmark]
  p-value  : p-value for Welch's t-test against baseline (H1: True difference in means > 0.3 ns)
  1 ns     : 1 Nanosecond (0.000000001 sec)
```

This benchmark was executed several times on 3 different machines (Linux/Windows/macOS with different hardware configurations). All the time I have the same conclusion based on the Welch's t-test: we can't detect statistically significant difference which is more than 0.3ns between mean values of `c8e3ba6c09` and `corefx#32389` for `Take2`. There are the density and timeline plots for this case:
![consoleapp11 takeoverheadbenchmark-take2-density](https://user-images.githubusercontent.com/2259237/46069346-0c3d7c00-c184-11e8-8987-0785e9795903.png)
![consoleapp11 takeoverheadbenchmark-take2-facettimelinesmooth](https://user-images.githubusercontent.com/2259237/46069407-2bd4a480-c184-11e8-8b45-a8f6c24ee8aa.png)
